### PR TITLE
IGNITE-21253 Implement a counter for number of rebalancing tables inside the zone

### PR DESCRIPTION
**Pull request by alievmirza**
https://issues.apache.org/jira/browse/IGNITE-21253

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 1]
```diff
@@ -432,7 +440,8 @@ public TableManager(
             HybridTimestampTracker observableTimestampTracker,
             PlacementDriver placementDriver,
             Supplier<IgniteSql> sql,
-            FailureProcessor failureProcessor
+            FailureProcessor failureProcessor,
+            ScheduledExecutorService rebalanceScheduler
```
Comment (kgusakov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:444]: javadoc must be updated

**Review by kgusakov — CHANGES_REQUESTED** [review]

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceRaftGroupEventsListener.java [hunk 2]
```diff
@@ -413,6 +510,14 @@ private void doOnNewPeersConfigurationApplied(PeersAndLearners configuration) {
                 retryPreconditions = and(retryPreconditions, con5);
             }
 
+            retryPreconditions = and(
+                    or(
+                            notExists(stableChangeTriggerKey(tablePartitionId)),
+                            value(stableChangeTriggerKey(tablePartitionId)).lt(ByteUtils.longToBytes(revision))
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceRaftGroupEventsListener.java]: Added

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 1 above)
Comment (alievmirza) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:453]: done

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngine.java [hunk 3]
```diff
@@ -214,6 +242,85 @@ public void onError(Throwable e) {
...
+
+                            for (CatalogTableDescriptor table : tables) {
+                                TablePartitionId replicaGrpId = new TablePartitionId(table.id(), partId);
+                                partitionTablesKeys.put(raftConfigurationAppliedKey(replicaGrpId), replicaGrpId);
+                            }
+
+                            Map<ByteArray, Entry> entriesMap = metaStorageManager.getAll(partitionTablesKeys.keySet()).get();
+
+                            entriesMap.forEach((key, stable) -> {
+                                doOnNewPeersConfigurationApplied(
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngine.java]: I've added check for revision and we skipp outdared events

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 4]
```diff
@@ -1778,9 +1765,7 @@ private CompletableFuture<Void> handleChangePendingAssignmentEvent(
                         // Table can be null only recovery, because we use a revision from the future. See comment inside
                         // performRebalanceOnRecovery.
                         if (table == null) {
-                            if (LOG.isInfoEnabled()) {
-                                LOG.info("Skipping Pending Assignments update, because table {} does not exist", tblId);
-                            }
+                            assert false : "Table cannot be null.";
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java]: reverted

**Review by kgusakov — APPROVED** [review]

**Pull request merged**